cupyx/scipy/interpolate/_bspline.py

```diff
@@ -0,0 +1,912 @@
+    nu : int
+        Order of derivative to evaluate.
+    extrapolate : int, optional
+        Whether to extrapolate to ouf-of-bounds points, or to return NaNs.
+    out : ndarray, shape (s, m)
+        Computed values of the spline at each of the input points.
+        This argument is modified in-place.
+    """
+    n = t.shape[0] - k - 1
+    intervals = cupy.empty_like(xp, dtype=cupy.int_)
+
+    # Compute intervals for each value
+    interval_kernel = _get_module_func(INTERVAL_MODULE, 'find_interval', xp)
+    interval_kernel((max(1, xp.shape[0] // 128),), (min(128, xp.shape[0]),),
+                    (t, xp, intervals, k, n, extrapolate))
+
+    # Compute interpolation
+    num_c = int(np.prod(c.shape[1:]))
+    temp = cupy.empty(xp.shape[0] * (2 * k + 1))
+    d_boor_kernel = _get_module_func(D_BOOR_MODULE, 'd_boor', xp)
+    d_boor_kernel((max(1, xp.shape[0] // 128),), (min(128, xp.shape[0]),),
+                  (t, c, k, nu, xp, intervals, out, temp, num_c, 1))
+
+
+def _make_design_matrix(x, t, k, extrapolate, indices):
+    """
+    Returns a design matrix in CSR format.
+    Note that only indices is passed, but not indptr because indptr is already
+    precomputed in the calling Python function design_matrix.
+
+    Parameters
+    ----------
+    x : array_like, shape (n,)
+        Points to evaluate the spline at.
+    t : array_like, shape (nt,)
+        Sorted 1D array of knots.
+    k : int
+        B-spline degree.
+    extrapolate : bool, optional
+        Whether to extrapolate to ouf-of-bounds points.
+    indices : ndarray, shape (n * (k + 1),)
+        Preallocated indices of the final CSR array.
+    Returns
+    -------
+    data
+        The data array of a CSR array of the b-spline design matrix.
+        In each row all the basis elements are evaluated at the certain point
+        (first row - x[0], ..., last row - x[-1]).
+
+    indices
+        The indices array of a CSR array of the b-spline design matrix.
+    """
+    n = t.shape[0] - k - 1
+    intervals = cupy.empty_like(x, dtype=cupy.int_)
+
+    # Compute intervals for each value
+    interval_kernel = _get_module_func(INTERVAL_MODULE, 'find_interval', x)
+    interval_kernel((max(1, x.shape[0] // 128),), (min(128, x.shape[0]),),
+                    (t, x, intervals, k, n, extrapolate))
+
+    # Compute interpolation
+    bspline_basis = cupy.empty(x.shape[0] * (2 * k + 1))
+    d_boor_kernel = _get_module_func(D_BOOR_MODULE, 'd_boor', x)
+    d_boor_kernel((max(1, x.shape[0] // 128),), (min(128, x.shape[0]),),
+                  (t, None, k, 0, x, intervals, None, bspline_basis, 0, 0))
+
+    data = cupy.zeros(x.shape[0] * (k + 1), dtype=cupy.float_)
+    design_mat_kernel = _get_module_func(
+        DESIGN_MAT_MODULE, 'compute_design_matrix', x, indices)
+    design_mat_kernel((max(1, x.shape[0] // 128),), (min(128, x.shape[0]),),
+                      (k, intervals, bspline_basis, data, indices))
+
+    return data, indices
```

Comment:
Note that if the structure inherits the scipy version, which minimizes the amount of stuff being sent between python and cython. ISTM the code will be easier to understand if indptr is constructed here as well.

See however a separate comment on this whole function.